### ISSUE TITLE
GamePhysics | Output Data Constraints aren't constrained

### ISSUE THREAD

**Issue by samm82** [JacquesCarette/Drasil#1283]
The Output Data Constraints table in GamePhysics lacks a column for the Physical Constraints, since the values aren't given any constraints in the Drasil code.

![outputDataConstraints](https://user-images.githubusercontent.com/35857611/57330505-3d9c0400-70e4-11e9-8070-de3fb5a2de13.png)

`posCons`, `veloCons`, `orientCons`, and `angVeloCons` are the relevant quantities.
https://github.com/JacquesCarette/Drasil/blob/5dee7ef7434683355d438fc1c256d268b5c4eb3c/code/drasil-example/Drasil/GamePhysics/Unitals.hs#L276-L293

Not sure who's working on this example @smiths so feel free to reassign.

**Comment by smiths** [comment]
Thank you @samm82.  Currently @oluowoj is working on improving the Game Physics example.  It needs some attention to get it up to the standard of the other examples.  I'll assign this to @oluowoj, but she might also reassign it, possibly to you.  :smile:

**Comment by oluowoj** [comment]
This is on my to-do list, I am not there yet, but will take a look.

**Comment by oluowoj** [comment]
@samm82 , working on this one at the moment, I thought it was going to be a quick one when we chatted about it yesterday...lol. It should be though, maybe I am just missing something.

I updated unitals.hs for the physical constraints column, I put in dummy figures just to see the column first,  but I still don't see the changes I am looking for. Any clues?
I have pushed my changes to 'outputConstraint' branch.

**Comment by samm82** [comment]
I pushed a commit to the branch that added the constraint to the placeholder variable for orientation, and the constraints column now displays. Based on the comment, it shouldn't be necessary because the "space is radians", but I'm not sure how that would enforced.

The reason you didn't see any changes is because you essentially didn't make any - you just "renamed" the pre-existing concepts, which doesn't change the external behaviour of Drasil. Adding the constraint changed the actual behaviour, as the quantity has more information.

It's interesting that if there aren't any constraints the whole column isn't displayed, but as soon as there is, the column is displayed (and empty cells are filled with "--").

**Comment by oluowoj** [comment]
Well I had to rename the output constraints because it was throwing some kind of error related to duplication since the constraints appear in both input and output.  Weird how the column display depends on a value. Thanks

**Comment by oluowoj** [comment]
@ Smith, there is a comment in the code that says --- constraints should not be necessary for angular orientation because space is radians----

Is that the case? See below what the constraints look like in the manual version

<img width="191" alt="image" src="https://user-images.githubusercontent.com/43192745/59148384-6ad71d00-89d6-11e9-9125-7d70e671fc06.png">

**Comment by JacquesCarette** [comment]
Your tag of @smiths didn't work (it should show in bold; the hint is that there is a space between the `@` and the name).

**Comment by smiths** [comment]
@oluowoj, the manual version is technically wrong.  Requiring the value be between 0 and 2 pi isn't a real physical constraint.  The angle can be any positive value; a value greater than 2 pi just means you would be going "around" more than once.  Any value greater than 2 pi can always be converted to an equivalent angle between 0 and 2 pi, but larger values are allowed by physics.  We can also have negative angles; that just means you "spin" in the opposite direction.

What I would like is if we had no physical constraint on phi, but then had a software constraint that phi lie between 0 and 2 pi.  I notice from the above that game physics doesn't have software constraints, but it should.

**Comment by JacquesCarette** [comment]
Did you ever get around to this @oluowoj ?

**Comment by oluowoj** [comment]
I don't think this is a 100% complete. The constraints were added but maybe it just beed to be double checked to make sure.

**Comment by JacquesCarette** [comment]
If you could please make sure, that would be great. If there are some things that still need done, please document them here, we can have one of the summer students deal with them.

**Comment by oluowoj** [comment]
Added to my to do list

**Comment by oluowoj** [comment]
What's left to be done in this one is:

1. Add a new column to the input data constraint table, column name: Software Constraints
2. Move the value in Physical constraints column of 'Input Data Constraints' table for phi to the software constraint column
3. Remove the value in Physical Constraint column of "Output Data Constraints" table, no value needed, just put a "--" - (most likely that the table will not display if there are no values for any of the parameters in the Physical constraints section"

**Comment by smiths** [comment]
@oluowoj, thank you for the analysis of what needs to be done to close this issue.  At the moment the issue is still assigned to you.  Will you be completing the tasks you outlined, or should we assign it to one of the summer students?

**Comment by oluowoj** [comment]
Would be nice if one of the summer students can take it up as per @JacquesCarette

**Comment by smiths** [comment]
@muhammadaliog3, I'll assign this issue to you.  The tasks that need to be done to close this issue are given in the comment by @oluowoj above.  Please contact @oluowoj if you have any questions.

**Comment by muhammadaliog3** [comment]
@oluowoj  I was able to complete the first 2 very easily, but I am struggling with the third one. As you said the physical constraint column will not display if no constraint is specified. Hence my first solution was to display the constraint of at least one physical constraint in one of the outputs. However, Constraints can only be one of the following: 
```haskell
data ConstraintReason = Physical | Software
data Constraint where
  Range          :: ConstraintReason -> RealInterval Expr Expr -> Constraint
  EnumeratedReal :: ConstraintReason -> [Double]               -> Constraint
  EnumeratedStr  :: ConstraintReason -> [String]               -> Constraint
```
So there is no easy implementation of an "empty constraint" among current constraints. If I add a new "empty constraint" to the Constraint datatype then I will have to make changes to the modules in the Imperative directory, which might influence code generation. 

My other solution is to change the way we generate all output tables by giving them a separate new function so that they show the physical constraint column regardless of whether or not there are any physical constraints. 

Do you think I am overcomplicating things? Is there already a way to add empty columns in Drasil that you perhaps know about?

**Comment by oluowoj** [comment]
@muhammadaliog3 , thanks for the update. I don't know of a way to add an empty column. @smiths, do we really need to show a table if there are no values. Could we just add some text to that column for now indicating there are no output constraints?

**Comment by muhammadaliog3** [comment]
Here is what the table would look like. I think it looks a little awkward.
![gamephys_output_table](https://user-images.githubusercontent.com/64814624/83039242-ea0c9900-a00b-11ea-86af-86c491543f8c.png)

**Comment by oluowoj** [comment]
Yeah, not what we want to see :). There is no use displaying only the parameters without values.

**Comment by muhammadaliog3** [comment]
Hold on I am onto an easier solution, just give me some time.

**Comment by smiths** [comment]
@oluowoj and @muhammadaliog3, the table also has the purpose of summarizing the output values.  Removing it would make the list of outputs implicit, which is not our goal.  We want the documentation to clearly show what are inputs and what are outputs.  We want an explicit list of outputs.

There is also another explanation for your observation that there are no constraints.  The explanation involves an important "revelation", so I don't want to just type it.  I want you to think about it.  No output data constraints are given.  One explanation is that there are no data constraints.  What is another explanation?  Hint - the game physics example is a work in progress.  :smile:

**Comment by smiths** [comment]
@muhammadaliog3, if the easier solution involves removing the list of outputs, it isn't the right solution.  :smile:

**Comment by muhammadaliog3** [comment]
So here is the solution I was making changes to the function outDataConstTbl and in the mean time added the functions lenPhysC and lenSfwC. The original code is defined in Doclang.Sections.SpecificSystemDescription.hs, but that changes can also be viewed in the Issue_1283_gamephysics branch.
```haskell
outDataConstTbl :: (Quantity c, Constrained c) => [c] -> LabelledContent
outDataConstTbl qlst 
            | hasAtleastOneConstraint ==True = mkDataConstraintTable [(S "Var", map ch qlst),
              (titleize' physicalConstraint, map (\_-> S (show $ length $ (map fmtPhys qlst))) qlst),
              (titleize' softwareConstraint, map fmtSfwr qlst)] "OutDataConstraints" $
              S "Output Data Constraints"
            | otherwise = mkDataConstraintTable [(S "Var",[]),
              (titleize' physicalConstraint, map (\_ -> S "--") qlst)] "OutDataConstraints" $
              S "Output Data Constraints"
            where 
              hasAtleastOneConstraint= (lenPhysC qlst + lenSfwC qlst)>0

lenPhysC :: (Constrained c) => [c] -> Int
lenPhysC qlst= length $ filter isPhysC $ foldl1 (++) $ map (^. constraints) qlst
lenSfwC :: (Constrained c) => [c] -> Int
lenSfwC  qlst= length $ filter isSfwrC $ foldl1 (++) $ map (^. constraints) qlst
-- | formats physical constraints
fmtPhys :: (Constrained c, Quantity c) => c -> Sentence
fmtPhys c = foldConstraints c $ filter isPhysC (c ^. constraints)

-- | formats software constraints
fmtSfwr :: (Constrained c, Quantity c) => c -> Sentence
fmtSfwr c = foldConstraints c $ filter isSfwrC (c ^. constraints)
```
Explaining the function: If there is at least one software or physical constraint (the number of Sfwr and physical constraints is calculated by lenSfwC and lenPhyC respectively), then proceed as normal. Otherwise I create a dummy column with the same name as the physical constraints, but this time contains only sentences of "--". Unfortunately when I build the GamePhysics example the GlassBr_SRS.tex file is empty although the code compiled. 

I am quite sure that the error is in the lenPhysC and lenSfwC, but debugging this will probably take a few hours and hence will have to wait till another day. But I am posting this here to get feedback on the solution (that is am I on the right track) and for whoever that reads this to tell me if they see a glaringly obvious mistake/improvement that I should make.

**Comment by muhammadaliog3** [comment]
So here is the solution I was making changes to the function outDataConstTbl. The original code is defined in Doclang.Sections.SpecificSystemDescription.hs, but these changes can also be viewed in the Issue_1283_gamephysics branch.
```haskell
outDataConstTbl :: (Quantity c, Constrained c) => [c] -> LabelledContent
outDataConstTbl qlst 
            | hasAtleastOneConstraint ==True = mkDataConstraintTable [(S "Var", map ch qlst),
              (titleize' physicalConstraint, map fmtPhys qlst),
              (titleize' softwareConstraint, map fmtSfwr qlst)] "OutDataConstraints" $
              S "Output Data Constraints"
            | otherwise = mkDataConstraintTable [(S "Var",map ch qlst),
              dummyPhysCol] "OutDataConstraints" $
              S "Output Data Constraints"
            where 
              lenPhysC = length $ filter isPhysC $ foldl (++) [] $ map (^. constraints) qlst 
              lenSfwC  = length $ filter isSfwrC $ foldl (++) [] $ map (^. constraints) qlst
              dummyPhysCol = (titleize' physicalConstraint, map (\_ -> S "_") qlst)
              hasAtleastOneConstraint= (lenSfwC  + lenPhysC ) > 0

fmtPhys :: (Constrained c, Quantity c) => c -> Sentence
fmtPhys c = foldConstraints c $ filter isPhysC (c ^. constraints)

-- | formats software constraints
fmtSfwr :: (Constrained c, Quantity c) => c -> Sentence
fmtSfwr c = foldConstraints c $ filter isSfwrC (c ^. constraints)
```
Explaining the function: If there is at least one software or physical constraint (the number of Sfwr and physical constraints is calculated by lenSfwC and lenPhyC respectively), then proceed as normal. Otherwise I create a dummy column with the same name as the physical constraints, but this time contains only sentences of "--", and make a table with the dummy column. 

I am posting this here to get feedback on the solution (that is am I on the right track) and for advice on cleaning the code. I see a few potential spots for cleaning
1. A better way to check ">0"
2. Perhaps add lenPhysC and lenSfwC as functions in Constraints.hs as they could be potentially used for another task.

**Comment by JacquesCarette** [comment]
Definitely yes to adding `lenPhysC` and `lenSfwrC` to Constaints.hs.

I would much prefer if the constraints column was completely removed if there are none. The easiest way to do that would be to create a `mkDataTable` that is like `mkDataConstraintTable`, just with the last column removed.

**Comment by muhammadaliog3** [comment]
I think @smiths said to add a constraints column even if there are none.

**Comment by smiths** [comment]
@muhammadaliog3, I don't remember exactly what I said, but my current thinking is that there should nearly always be some quantities that have uncertainty.  Since this is the case, there would normally be an uncertainty column.  In the case where everything truly is 100% certain, I'm okay if the column does not appear.
